### DISCUSSION TITLE
jwt in the authorization header even if authorization failed

### DISCUSSION THREAD

**Discussion by opya** [jeremyevans/rodauth#342] [General]
Hi guys,

I wondered if there is a specific reason to `set_jwt` in `rereturn_json_response` even if the login failed. Or I misconfigured something and did something wrong.

https://github.com/jeremyevans/rodauth/blob/cbdfde53f4942eb5f2f52299a68223b9e2e62bf8/lib/rodauth/features/jwt.rb#L137-L140

![Screenshot from 2023-06-02 08-39-49](https://github.com/jeremyevans/rodauth/assets/3123068/106899a5-d89a-4e5e-a417-8f234b9e225e)

**Comment by jeremyevans** [comment]
As the code you posted shows, JWTs are set for all JSON responses if you are using the jwt feature.  Is this causing a problem in your application?

**Reply by thedumbtechguy** [reply]
@jeremyevans I saw a similar report on rodauth rails https://github.com/janko/rodauth-rails/discussions/213

That's what brought me here. Appears the returned token is valid.
And that might be a problem since the login attempt failed.

I'm yet to test myself but that's what the other report seems to claim.

**Reply by jeremyevans** [reply]
The response body and status code indicate whether a request was successful.  A JWT is always set for consistency, mirroring the non-JWT case where a session cookie is set. This is by design.  It is possible to override the behavior and not set JWT for errors:

```ruby
set_jwt_token do |token|
  super(token) unless json_response_error?
end
```

**Comment by opya** [comment]
no, no problems just curiosity on my side. I'm using jwt feature with `use_jwt? false` to fallback and use rack session and was wondering why the Authorization header appear. Thank you @jeremyevans

**Reply by janko** [reply]
Just wondering, why are you using the JWT feature if you're relying on Rack session? I would use just the JSON feature in this case.

**Reply by opya** [reply]
I'm using roda, sequel, and rodauth as the backend and react spa frontend. The react spa is an administrative app and there I want to use sessions instead of JWT tokens because I'm not sure which is the best way to store the token in the browser, so I'm trying to use sames_site: strict cookies. The ruby app is divided in tho two, the first one is that which interacts with react spa for administration and uses JWT as a cookie and the second one is API where we will use JWT as a token and then give the token to the clients in order to integrate with the second part of the app. I'm not sure which is the correct way to do the above with ruby/roda/sequel stack, I'm curious how you use roda and friends ?  :)

**Reply by janko** [reply]
Right, I'm the same, I don't know how to securely store the token in the browser, so I would prefer using Rack session for a SPA as well 🙂
